### PR TITLE
Fullcube/validates uniqueness of case insensitive

### LINKED ISSUE #607: Validate unique field case insesitive
Hello,

we have tried today to add a unique validation to a model of countries on the isoCode. I think there is missing a flag boolean "caseSensitive". Right now we can have 2 countries with iso codes: de, DE an unique validation doesn't trigger error.

### PR DESCRIPTION
**Pull request by beeman**
This is an improved version of this PR https://github.com/strongloop/loopback-datasource-juggler/pull/798

---

This patch adds an option to check the uniqueness of a value in a case insensitive way.

Should fix #607, though I'm not sure if I overlooked any edge cases.

### CODE REVIEW

**Comment by Amir-61** [comment]
Thanks @beeman for the PR. There are some Merge conflicts, could you please rebase your code on top of `master` and squash your commit into a single commit following [this standard](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html). Thanks!

**Comment by mrfelton** [comment]
@Amir-61 I've rebased this patch off master.

**Comment by Amir-61** [comment]
@mrfelton @beeman the commit message is not following the standards defined [here](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html); thats why the test failed. Could you please fix it.

**Comment by Amir-61** [comment]
you can use `git commit --amend` to change the commit.

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by mrfelton** [comment]
@Amir-61 I've adjusted the commit message.

**Comment by Amir-61** [comment]
@slnode test please

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by Amir-61** [comment]
This feature and the logic make sense to me; I added `needs-doc` label as well.

@beeman/ @mrfelton Can we please have specific test for conf `{ ignoreCase: false }` as well? 

@bajtos What is your thought about this feature?

**Comment by Amir-61** [comment]
@beeman @mrfelton Thanks for contribution 👍

File: test/validations.test.js [hunk 1]
```diff
@@ -525,6 +525,52 @@ describe('validations', function() {
...
+        u.save(function() {
+          var u2 = new User({ email: 'Me@My.com' });
+          u2.isValid(function(valid) {
+            valid.should.be.true;
+            done();
+          });
+        });
+      })).should.be.false;
+    });
+
```
Comment (Amir-61) [inline test/validations.test.js]: Nitpick: Please remove this extra blank line

File: test/validations.test.js [hunk 2]
```diff
@@ -525,6 +525,52 @@ describe('validations', function() {
         })).should.be.false;
       });
     });
+
+    it('should validate uniqueness case insensitive', function(done) {
```
Comment (Amir-61) [inline test/validations.test.js]: Please do not start test names with `should`, even though we have followed some bad practices in our previous code. Please see our [style-guide](https://github.com/strongloop/loopback-contributor-docs/blob/master/style-guide.md#test-naming)

Please fix it for other tests as well.

**Comment by bajtos** [comment]
@Amir-61 I think the feature request to have a case-insensitive unique constraint is a valid one.

The only concern I have is whether we have an operator that can be used in all connectors we support (SQL and NoSQL).

I'll leave further review to you @Amir-61 and @raymondfeng for final approval if needed.

**Comment by beeman** [comment]
Any updates on this PR? :)

**Comment by Amir-61** [comment]
>The only concern I have is whether we have an operator that can be used in all connectors we support (SQL and NoSQL).

Thats my concern as well... I'm not sure if all supported connectors have this functionality... @raymondfeng any thought on this?

**Comment by Amir-61** [comment]
@slnode test please

**Comment by beeman** [comment]
Hey @Amir-61, thanks for your reply. 

Can I do anything to make sure this works? Would it work to test against MySQL/Oracle manually or are there better ways to do this?

**Comment by Amir-61** [comment]
>Can I do anything to make sure this works? Would it work to test against MySQL/Oracle manually or are there better ways to do this?

AFICT tests in `test/validations.test.js` are not run by connectors, whereas for example tests in `test\persistentence-hook.suite.js` are run by each connectors (dependencies); hence connectors which are not running your added tests in `test/validations.test.js`. I believe the only way is to add test for each connector individually to make sure they are passing. @raymondfeng  @bajtos for more thought?

**Comment by Amir-61** [comment]
@slnode test please

**Comment by beeman** [comment]
@Amir-61 I've rebased the commit to the latest master, can you please consider merging or tell me what needs to be done to get it in? Thanks!

**Comment by Amir-61** [comment]
@beeman Thanks for the update. @loay is going to take care of that.

**Comment by loay** [comment]
Hi @beeman There are few linting errors in this pr. 
Please run this command `./node_modules/.bin/eslint --fix .` as it should fix these issues, and/or will tell you what you need to do to get it fixed. 

Also, please make sure you have addressed Amir previous comments. 
We will give it another round of review so we can merge it asap
Thanks for your contribution.

**Comment by loay** [comment]
@slnode test please

**Comment by beeman** [comment]
@Amir-61 @loay thanks!

I've fixed the linting errors and updated the test names based on the previous comments.

**Comment by loay** [comment]
@slnode test please

File: test/validations.test.js [hunk 3]
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
...
+        });
+      })).should.be.false;
+    });
+
+    it('passes case insensitive validation with string that needs escaping', function(done) {
+      User.validatesUniquenessOf('email', {ignoreCase: true});
+      var u = new User({email: 'me+me@my.com'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
+        u.save(function() {
```
Comment (bajtos) [inline test/validations.test.js]: `u.save` can fail with an error, please handle it.

```
u.save(function(err) {
  if (err) return done(err);
  // ...
});
```

File: test/validations.test.js [hunk 4]
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
...
+        });
+      })).should.be.false;
+    });
+
+    it('passes partial case insensitive validation with string that needs escaping', function(done) {
+      User.validatesUniquenessOf('email', {ignoreCase: true});
+      var u = new User({email: 'also+me@my.com'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
+        u.save(function() {
```
Comment (bajtos) [inline test/validations.test.js]: `u.save` can fail with an error, please handle it.

File: test/validations.test.js [hunk 5]
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
...
+          });
+        });
+      })).should.be.false;
+    });
+
+    it('passes case insensitive validation with string that needs escaping', function(done) {
+      User.validatesUniquenessOf('email', {ignoreCase: true});
+      var u = new User({email: 'me+me@my.com'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
```
Comment (bajtos) [inline test/validations.test.js]: https://github.com/moll/js-must#beware-of-libraries-that-assert-on-property-access

I believe `.true` is a function in recent versions of shouldjs.

```js
valid.should.be.true();
```

File: test/validations.test.js [hunk 6]
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
         })).should.be.false;
       });
     });
+
+    it('passes case insensitive validation', function(done) {
+      User.validatesUniquenessOf('email', {ignoreCase: true});
+      var u = new User({email: 'hey'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
+        u.save(function() {
```
Comment (bajtos) [inline test/validations.test.js]: `u.save` can fail with an error, please handle it.

**Comment by loay** [comment]
@bajtos I was also thinking of having at least one test with `ignoreCase: false` to validate the behavior. What do you think?

**Comment by beeman** [comment]
@bajtos thanks for reviewing, I've addressed your comments.

@loay good point, I've added all 3 tests with `ignoreCase` set to `false`.

Please let me know if there's anything else to do to land this.

**Comment by loay** [comment]
@slnode test please

**Comment by loay** [comment]
@beeman thanks .. let's run the test one more time to ensure everything is fine on CI before merging

**Comment by loay** [comment]
@beeman can you please rebase? Thanks.

**Comment by beeman** [comment]
@loay Rebase done 👍

**Comment by loay** [comment]
@slnode test please

**Comment by bajtos** [comment]
As I was re-reading the discussion, https://github.com/strongloop/loopback-datasource-juggler/pull/818#issuecomment-256737192 caught my attention.

> > Can I do anything to make sure this works? Would it work to test against MySQL/Oracle manually or are there better ways to do this?

> AFICT tests in `test/validations.test.js` are not run by connectors, whereas for example tests in `test\persistentence-hook.suite.js` are run by each connectors (dependencies); hence connectors are not running your added tests in `test/validations.test.js`. I believe the only way to verify is to add tests for each connector individually to make sure they are passing. @raymondfeng  @bajtos for more thought?

I think most connectors are running `test/common.batch.js`. It should be enough to add `require('test/validations.test.js')` to that file in order to start running validation tests in each connector. However, I think this should be done in a standalone pull request, because we may need to fix some of the validation tests to avoid assumptions that don't hold for all downstream connectors.

**Comment by loay** [comment]
@beeman Thanks for your contribution.

**Comment by beeman** [comment]
Thanks for your input and help getting it landed guys 👍🏼

**Pull request merged**